To use certificate in Web App (#467)
Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2: 4-WebApp-your-API/4-1-MyOrg/README-use-certificate.md

# Web app using client certificates with Microsoft.Identity.Web

> There are different approaches to [use certificates with Microsoft.Identity.Web](https://github.com/AzureAD/microsoft-identity-web/wiki/Certificates). In this Readme, we will explore use of certificate from a path on the disk.

Reviewer: ```suggestion
There are different approaches for [using certificates with Microsoft.Identity.Web](https://github.com/AzureAD/microsoft-identity-web/wiki/Certificates). In this Readme, we will explore use of certificate from a path on the disk.
```

## Use Self Signed Certificate from a path

> To use certificates instead of an application secret you will need to do little changes to what you have done so far:

Reviewer 2: ```suggestion
To use certificates instead of an application secret you will need to do **some** changes to what you have done so far:
```

Reviewer: "make some changes"

- generate a certificate and export it, if you don't have one already
- register the certificate with your application in the Azure AD portal
- Update appsettings.json

### Generate a certificate

> #### On Windows

Reviewer 2: Instead of "On Windows" vs "On other OSs", it might be better to say "with Powershell" vs. "with OpenSSL"

> To complete this step, you will use the `New-SelfSignedCertificate` Powershell command. You can find more information about the New-SelfSignedCertificat command [here](https://docs.microsoft.com/powershell/module/pkiclient/new-selfsignedcertificate).

Reviewer 2: ```suggestion
To complete this step, you will use the `New-SelfSignedCertificate` Powershell command. You can find more information about the New-**SelfSignedCertificate** command [here](https://docs.microsoft.com/powershell/module/pkiclient/new-selfsignedcertificate).
```

1. Open PowerShell and run New-SelfSignedCertificate with the following parameters to create a self-signed certificate in the user certificate store on your computer:

    ```PowerShell
    $cert=New-SelfSignedCertificate -Subject "CN=WebAppCert" -CertStoreLocation "Cert:\CurrentUser\My"  -KeyExportPolicy Exportable -KeySpec Signature
    ```

1. Export this certificate using the "Manage User Certificate" MMC snap-in accessible from the Windows Control Panel. You can also add other options to generate the certificate in a different
store such as the Computer or service store (See [How to: View Certificates with the MMC Snap-in](https://docs.microsoft.com/dotnet/framework/wcf/feature-details/how-to-view-certificates-with-the-mmc-snap-in)).

Alternatively you can use an existing certificate if you have one (just be sure to record its name for the next steps).

Export one with private key as WebAppCert.pfx and another as WebAppCert.cer without private key.

#### On other Operating Systems

> Build and install **OpenSSL** for your **OS** following the guide at [github.com/openssl](https://github.com/openssl/openssl#build-and-install). If you like to skip building and get a binary distributable from the community instead, check the [OpenSSL Wiki: Binaries](https://wiki.openssl.org/index.php/Binaries) page.

Reviewer: We shouldn't need to build and install.
We should tell them to use it.

Reviewer: (because I think it comes bundled with MacOS and most Linux distros)

> Afterwards, add the path to `OpenSSL` to your **environment variables** so that you can call it from anywhere.

Reviewer: this is also unnecessary by default in MacOS and most Linux distros.

Type the following in a terminal. You will be prompted to enter some identifying information and a *passphrase*:

```console

>     openssl req -x509 -newkey rsa:4096 -sha256 -days 365 -keyout example.key -out WebAppCert.cer -subj "/CN=example.com" -addext "subjectAltName=DNS:example.com"

Reviewer: ```suggestion
    openssl req -x509 -newkey rsa:4096 -sha256 -days 365 -keyout example.key -out example.cer -subj "/CN=example.com" -addext "subjectAltName=DNS:example.com"
```
as per dogan's suggestion below, we should also change this line for matching key+cert filenames.

Author: updated to use webapp.
Thanks.

    Generating a RSA private key
    ...........................................................................................................................................................................................................................................................++++
    ......................................................................................................++++
    writing new private key to 'example.key'
    Enter PEM pass phrase:
    Verifying - Enter PEM pass phrase:
    ----- 
```

> After that, the following files should be generated: `example.key`, `WebAppCert.cer`.

Reviewer 2: ```suggestion
After that, the following files should be generated: `example.key`, `WebAppCert.cer`.
```
should have the same name for consistency

> After that, the following files should be generated: `example.key`, `WebAppCert.cer`.

Reviewer 2: If you need a .pfx file, you can combine .key and .cer with this command:

`openssl pkcs12 -export -out webapp.pfx -inkey webapp.key -in webapp.crt`

Author: Added the command.
Thanks.

### Add the certificate for the web application in Azure AD

> 1. Navigate back to to the [Azure portal](https://portal.azure.com).

Reviewer 2: ```suggestion
1. Navigate back to the [Azure portal](https://portal.azure.com).
```

In the left-hand navigation pane, select the **Azure Active Directory** service, and then select **App registrations**.

> 1. In the resultant screen, select the `TodoListClient-aspnetcore-webapi` application.

Author: It's the client app, here is the section to [Register the client app (TodoListClient-aspnetcore-webapi)](
 https://github.com/Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2/blob/Shama-K/Updates/4-WebApp-your-API/4-1-MyOrg/README.md#register-the-client-app-todolistclient-aspnetcore-webapi)

1. In the **Certificates & secrets** tab, go to **Certificates** section:
1. Click on **Upload certificate** and, in click the browse button on the right to select the certificate you just exported, WebAppCert.cer (or your existing certificate).
1. Click **Add**.

### Configure the Visual Studio project

To change the visual studio project to enable certificates you need to:

1. Open the `Client\appsettings.json` file
1. Find the app key `ClientCertificates` and add the keys as displayed below:

    "ClientCertificates": [
      {
        "SourceType": "",
        "CertificateDiskPath": "",
        "CertificatePassword": ""
      }
    ]

    Update values of the keys:

    `SourceType` to `Path`.

>     `CertificateDiskPath` to the path where certificate exported with private key (WebAppCert.pfx) is stored.

Reviewer 2: ```suggestion
    `CertificateDiskPath` to the path where certificate exported with private key (WebAppCert.pfx) is stored e.g. `C:\\Users\\Jane\\Desktop\\webapp.pfx`
```

# Web app using client certificates with Microsoft.Identity.Web

There are different approaches to [use certificates with Microsoft.Identity.Web](https://github.com/AzureAD/microsoft-identity-web/wiki/Certificates). In this Readme, we will explore use of certificate from a path on the disk.

## Use Self Signed Certificate from a path

To use certificates instead of an application secret you will need to do some changes to what you have done so far:

- generate a certificate and export it, if you don't have one already
- register the certificate with your application in the Azure AD portal
- Update appsettings.json

### Generate a Self Signed certificate

#### with Powershell

To complete this step, you will use the `New-SelfSignedCertificate` Powershell command. You can find more information about the New-SelfSignedCertificate command [here](https://docs.microsoft.com/powershell/module/pkiclient/new-selfsignedcertificate).

1. Open PowerShell and run New-SelfSignedCertificate with the following parameters to create a self-signed certificate in the user certificate store on your computer:

    ```PowerShell

>     $cert=New-SelfSignedCertificate -Subject "CN=WebAppCert" -CertStoreLocation "Cert:\CurrentUser\My"  -KeyExportPolicy Exportable -KeySpec Signature

Reviewer: ```suggestion
    $cert=New-SelfSignedCertificate -Subject "CN=webapp" -CertStoreLocation "Cert:\CurrentUser\My"  -KeyExportPolicy Exportable -KeySpec Signature
```

    ```

1. Export this certificate using the "Manage User Certificate" MMC snap-in accessible from the Windows Control Panel. You can also add other options to generate the certificate in a different
store such as the Computer or service store (See [How to: View Certificates with the MMC Snap-in](https://docs.microsoft.com/dotnet/framework/wcf/feature-details/how-to-view-certificates-with-the-mmc-snap-in)).

Alternatively you can use an existing certificate if you have one (just be sure to record its name for the next steps).

> Export one with private key as WebAppCert.pfx and another as WebAppCert.cer without private key.

Reviewer: ```suggestion
Export one with private key as webapp.pfx and another as webapp.cer without private key.
```

#### with OpenSSL

Type the following in a terminal. You will be prompted to enter some identifying information and a *passphrase*:

```console

>     openssl req -x509 -newkey rsa:4096 -sha256 -days 365 -keyout webapp.key -out webapp.cer -subj "/CN=webapp.com" -addext "subjectAltName=DNS:webapp.com"

Reviewer: ```suggestion
    openssl req -x509 -newkey rsa:4096 -sha256 -days 365 -keyout webapp.key -out webapp.cer -nodes -batch
```
putting this here will do the following:
-nodes (means `no DES`) : no need for a passphrase for DES encryption
-batch : don't prompt for any attributes like common name, dns, etc. This cert doesn't need these.

    Generating a RSA private key
    ...........................................................................................................................................................................................................................................................++++
    ......................................................................................................++++
    writing new private key to 'webapp.key'
    Enter PEM pass phrase:
    Verifying - Enter PEM pass phrase:
    ----- 
```

Generate the webapp.pfx certificate with below command:

```console
  openssl pkcs12 -export -out webapp.pfx -inkey webapp.key -in webapp.cer
```

> After that, the following files should be generated: `webapp.key`, `webapp.cer` and `webapp.pfx`.

Reviewer: ```suggestion
Enter an export password when prompted and make a note of it.
The following files should be generated: `webapp.key`, `webapp.cer` and `webapp.pfx`.
```

### Add the certificate for the web application in Azure AD

1. Navigate back to the [Azure portal](https://portal.azure.com).
In the left-hand navigation pane, select the **Azure Active Directory** service, and then select **App registrations**.
1. In the resultant screen, select the `TodoListClient-aspnetcore-webapi` application.
1. In the **Certificates & secrets** tab, go to **Certificates** section:

> 1. Click on **Upload certificate** and, in click the browse button on the right to select the certificate you just exported, WebAppCert.cer (or your existing certificate).

Reviewer: ```suggestion
1. Click on **Upload certificate** and, in click the browse button on the right to select the certificate you just exported, example.cer (or your existing certificate).
```
this one too

1. Click **Add**.

### Configure the Visual Studio project

To change the visual studio project to enable certificates you need to:

1. Open the `Client\appsettings.json` file
1. Find the app key `ClientCertificates` and add the keys as displayed below:

>     "ClientCertificates": [
>       {
>         "SourceType": "",
>         "CertificateDiskPath": "",
>         "CertificatePassword": ""
>       }
>     ]

Reviewer: ```suggestion
    "ClientSecret": ""
    "ClientCertificates": [
      {
        "SourceType": "Path",
        "CertificateDiskPath": "webapp.pfx",
        "CertificatePassword": "the export password you recorded in the previous step"
      }
    ]
```

Author: Examples for the values are specified in `Update values of the keys` section.

    Update values of the keys:

    `SourceType` to `Path`.

>     `CertificateDiskPath` to the path where certificate exported with private key (WebAppCert.pfx) is stored.

Reviewer: ```suggestion
    `CertificateDiskPath` to the path where certificate exported with private key (webapp.pfx) is stored.
```
